.github/workflows/deploy.yml

```diff
@@ -1,52 +1,69 @@
-name: Deploy mdBook to GitHub Pages
+name: Build and Deploy mdBook
+# Tool versions are managed in versions.toml at the repository root.
+# To update mdbook or plugin versions, edit versions.toml and the CI will automatically use them.
 
 on:
-  # Runs on pushes targeting the default branch
   push:
     branches: ["main"]
+  
+  pull_request:
+    branches: ["main"]
 
-  # Allows you to run this workflow manually from the Actions tab
   workflow_dispatch:
 
-# Sets permissions of the GITHUB_TOKEN to allow deployment to GitHub Pages
 permissions:
   contents: read
   pages: write
   id-token: write
 
-# Allow only one concurrent deployment, skipping runs queued between the run in-progress and latest queued.
-# However, do NOT cancel in-progress runs as we want to allow these production deployments to complete.
 concurrency:
   group: "pages"
   cancel-in-progress: false
 
 jobs:
-  # Build job
   build:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
-        uses: actions/checkout@v4
+        uses: actions/checkout@v6
+      
+      # Cache Rust toolchain and cargo registry
+      - name: Cache Rust
+        uses: actions/cache@v4
+        with:
+          path: |
+            ~/.cargo/bin/
+            ~/.cargo/registry/index/
+            ~/.cargo/registry/cache/
+            ~/.cargo/git/db/
+            ~/.cargo/.crates.toml
+            ~/.cargo/.crates2.json
+          key: ${{ runner.os }}-cargo-${{ hashFiles('versions.toml') }}
+          restore-keys: |
+            ${{ runner.os }}-cargo-
           
-      - name: Install mdBook
-        run: cargo install mdbook
+      - name: Install mdBook and plugins
+        run: |
+          chmod +x scripts/install-tools.sh
+          ./scripts/install-tools.sh
+      
+      - name: Build with mdBook
+        run: mdbook build ./better-code
         
+      # Only deploy from main branch (push or manual dispatch)
       - name: Setup Pages
+        if: (github.event_name == 'push' || github.event_name == 'workflow_dispatch') && github.ref == 'refs/heads/main'
         id: pages
         uses: actions/configure-pages@v5
         
-      - name: Build with mdBook
-        run: |
-          cd better-code
-          mdbook build
-          
       - name: Upload artifact
-        uses: actions/upload-pages-artifact@v3
+        if: (github.event_name == 'push' || github.event_name == 'workflow_dispatch') && github.ref == 'refs/heads/main'
+        uses: actions/upload-pages-artifact@v4
         with:
           path: ./better-code/book
 
-  # Deployment job
   deploy:
+    if: (github.event_name == 'push' || github.event_name == 'workflow_dispatch') && github.ref == 'refs/heads/main'
```

Comment:
Would it be better to have a dependent workflow for deploy that only runs on push to main?

Reply:
Yes - updated.